refactor(streams): Refactor response stream handling.

haadcode opened the pull request:
This PR will:

- Refactor the response stream handling.
- Add dagnode-stream to transform file results to DAGNode objects.
- Add stream-to-value to convert a response stream to a single value.
- Refactor request-api so that chunked JSON objects are not buffered anymore. This touches a bunch of the files pushing the transform function down to the individual commands.

The general idea in this PR is to  cleanup the architecture and implementation regards to how the http responses are handled. Previously some of it was done in request-api.js and some in the individual commands making request-api coupled with the commands' response type. I've moved the response transform logic down to the individual commands so that request-api is agnostic to what is returned to the consumer of the commands. I've also refactored individual types of transformations to their own files to make it clear, clean and much more reusable (we can consider moving them to their own modules, but not in this PR).

dignifiedquire commented on src/dagnode-stream.js:
```diff
@@ -0,0 +1,55 @@
+    this._send = opts.send
+  }
+
+  static streamToValue (send, inputStream, callback) {
```
is `static` available in Node4?

haadcode replied:
Is it? Is it not? And if so, why is that important. Provide context and solutions or guidance.

diasdavid commented on src/api/add.js:
```diff
@@ -1,24 +1,26 @@
 
-    if (!good) {
-      callback(new Error('"files" must be a buffer, readable stream, or array of objects'))
+    if (!ok) {
```
unnecessary change

diasdavid replied:
fine :)

diasdavid commented on src/api/add.js:
```diff
@@ -1,24 +1,26 @@
-    if (!good) {
-      callback(new Error('"files" must be a buffer, readable stream, or array of objects'))
+    if (!ok) {
+      return callback(new Error('"files" must be a buffer, readable stream, or array of objects'))
```
good catch

diasdavid commented on src/tar-stream-to-objects.js:
```diff
@@ -1,38 +1,45 @@
+    callback(null, outputStream)
+  }
+
+  _read () {}
```
A Class with a constructor and a static method? What is the goal here?

diasdavid commented on src/dagnode-stream.js:
```diff
@@ -0,0 +1,55 @@
+  Output object format:
+  {
+    path: '/path/to/file/foo.txt',
+    hash: 'Qma4hjFTnCasJ8PVp3mZbZK5g2vGDT4LByLJ7m8ciyRFZP',
```
hash or `multihash`?

diasdavid replied:
This "Output object format" is not a DAGNode `Transforms a stream of objects to DAGNodes and outputs them as objects.`, what am I missing?

diasdavid replied:
🛎

haadcode replied:
Changed the description. Does that make sense now?

haadcode replied:
Here: https://github.com/ipfs/js-ipfs-api/pull/465/commits/1566751ee3c703dfc18a5be59bcf8f5d4d48a580

haadcode replied:
@diasdavid I'm not sure about the `multihash` vs. `hash`. It's a breaking change again and I'd be hesitant to change it in this PR. Have we made the change elsewhere in the code already? Does anything that returns DAGNode-like information return .multihash instead of .hash? If we have, I can change it so that we're consistent, but if not, I would leave it for later.

diasdavid commented on src/request-api.js:
```diff
@@ -173,33 +144,21 @@ exports = module.exports = function getRequestAPI (config) {
 
-  // Wraps the 'send' function such that an asynchronous
-  // transform may be applied to its result before
-  // passing it on to either its callback or promise.
```
can you preserve the comment please?

haadcode replied:
This comment doesn't make sense, and it doesn't hold true for the new implementation.

diasdavid replied:
It was trying to convey that before the response is returned, either through the callback or promise, we transform the response from go-ipfs to match interface-ipfs-core definition.

haadcode replied:
Added the description back and trying to describe as simple as I can what it does. What do you think?

haadcode replied:
Here: https://github.com/ipfs/js-ipfs-api/pull/465/commits/1566751ee3c703dfc18a5be59bcf8f5d4d48a580

diasdavid replied:
👍

diasdavid commented on src/stream-to-value.js:
```diff
@@ -0,0 +1,12 @@
+  res.pipe(concat((data) => callback(null, data)))
+}
+
+module.exports = streamToValue
```
In case of error on the stream (connection dropped) this fails silently

diasdavid commented on src/stringlist-to-array.js:
```diff
@@ -0,0 +1,9 @@
+  cb(null, res.Strings || [])
+}
+
+module.exports = stringlistToArray
```
What is a string list? Is there a better name in this context?  Does it has to be a separate file to do this?

diasdavid replied:
@whyrusleeping wanna comment?

haadcode replied:
We shouldn't change this on go-ipfs side for this PR. If we do that, it's an API breaking change that we need to handle with proper care. That's why I put it in one file, so that it's wrapped and isolated and if/once we change the return values, we only need to change change it in this file and whichever command is using the current API doesn't need changes. Obviously once you do that, we don't need the stringlist-to-array.js conversion function and can get rid of the file completely, but I recommend we keep it as such now so that we isolate the data conversion to one place. Does that make sense?

diasdavid replied:
Understood :)

haadcode commented:
Thanks for the review @dignifiedquire and @diasdavid. Will work on this soon, mainly change it to use pump as well as address other CR pieces.

diasdavid commented on src/dagnode-stream.js:
```diff
@@ -0,0 +1,55 @@
+        path: obj.Name,
+        hash: obj.Hash,
+        size: node.size
+      }
```
Not really a `dag`, it is just a json version.

diasdavid commented on src/tar-stream-to-objects.js:
```diff
@@ -1,38 +1,45 @@
+    let outputStream = new TarStreamToObjects()
+
+    inputStream
+      .pipe(tar.extract())
```
better to use `pump`

haadcode commented:
Will update this today and address the CR feedback.

haadcode commented:
I've refactored the pipes to use pump. I wasn't sure if the pump is needed in dagnode-stream.js since streamToValue internally uses pump (so I take it would handle the errors), but I added it just in case. The tests are not running in the browser atm, but run fully on node (will look into why the browser tests are not running).

I also reworked the comments for the two functions @diasdavid had a confusion about, hope they're now more descriptive. I refactored the TarStreamToObjects according to @diasdavid's question, perhaps now the intent is clearer. I couldn't get the ReadableStream to work when created directly with `new Readable() and adding `_read() = () => {}` function to it like it's normally done, but by wrapping it in a class it works. If you have an idea how to make it work without the class, please do refactor it.

There's also a problem with browser version now as we're using pump because its dependency on `fs`. @dignifiedquire suggested we fix it there: https://github.com/mafintosh/pump/issues/6. If you know how, please do.

So while this is still WIP, can you take a look and see if there's anything else to address in this one?

dignifiedquire commented:
@haadcode pump fix is here: https://github.com/mafintosh/pump/pull/13 and the issue with cors was correct, you changed them, now they are back: https://github.com/ipfs/js-ipfs-api/pull/465/commits/b2c09b2a40cc2761dc767abb0a6400d1c24df859

haadcode commented:
Should we wait for the PR to get merged or can we use the PR commit as our dep in the meanwhile?

> the issue with cors was correct, you changed them, now they are back: b2c09b2

Interesting. I had to change it in the first place because it wasn't working. Good that it's fixed now 👍

dignifiedquire commented:
> Should we wait for the PR to get merged or can we use the PR commit as our dep in the meanwhile?

Pinged @mafintosh on irc, if it's not merged when we want to ship this lets switch to the fork.

dignifiedquire commented on package.json:
```diff
@@ -43,6 +42,7 @@
     "peer-id": "^0.8.1",
     "peer-info": "^0.8.1",
     "promisify-es6": "^1.0.2",
+    "pump": "^1.0.1",
```
lets bump this to `1.0.2` so we are sure to have the fix in

dignifiedquire commented on src/dagnode-stream.js:
```diff
@@ -0,0 +1,59 @@
+
+  static streamToValue (send, inputStream, callback) {
+    const outputStream = pump(inputStream, new DAGNodeStream({ send: send }), (err) => {
+      if (err) callback(err)
```
@diasdavid will want this to be 

```js
if (err) {
  callback(err)
} 

haadcode commented:
I fixed the style issues, updated pump to 1.0.2 and rebased on master.

> Other than the hash vs multihash thing, it LGTM. I'm fine if Files API uses hash, although I would prefer consistency.

@diasdavid would you be ok if we leave the change from `.hash` to `.multihash` to a separate PR? This is a large PR in itself and I'm afraid of the consequences the aforementioned change would have on users (being an API change). Perhaps you have more confidence to do it as you know the codebase throughout, but I would prefer doing it in one, separate PR where we update *all* functions (that return `.hash` currently) at the same time. Let me know what you think.

diasdavid commented:
> @diasdavid would you be ok if we leave the change from .hash to .multihash to a separate PR? 

sounds good :)

dignifiedquire commented:
@diasdavid did you pull in the latest version of #465 before merging?

dignifiedquire commented:
Also this is using ipfsd-ctl fork from @haadcode I though we wanted to wait for that release and switch to happen

haadcode commented:
Didn't even notice this got merged. Thank you! 😄  🎉
